.github/workflows/commit-checker.yml

```diff
@@ -0,0 +1,42 @@
+name: Commit checker
+
+on:
+  pull_request:
+
+jobs:
+  commit-checker:
+    name: Commit checker
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout ${{ github.base_ref }} branch
+      uses: actions/checkout@v2
+      with:
+        ref: ${{ github.base_ref }}
+
+    - name: Switch to ${{ github.ref }} branch
+      run: |
+        git -c protocol.version=2 fetch --no-tags --prune --negotiation-tip=origin/${GITHUB_BASE_REF} origin +${GITHUB_SHA}:${GITHUB_REF}
+        git checkout --progress --force ${GITHUB_REF}
+        git log origin/${GITHUB_BASE_REF}..HEAD^2
+
+    - name: Checkout commit-checker
+      uses: actions/checkout@v2
+      with:
+        repository: OpenTTD/OpenTTD-git-hooks
+        path: git-hooks
+        ref: master
+
+    - name: Check commits
+      run: |
+        is_on_top_of_master=$(git branch $(git rev-parse --abbrev-ref HEAD) --contains origin/${GITHUB_BASE_REF} | wc -l)
+
+        if [ "${is_on_top_of_master}" -ne 1 ]; then
+            echo "ERROR: This commit is not on top of ${GITHUB_BASE_REF}"
+            echo "(no further validation is done due to this error)"
+            exit 1
+        fi
+        echo "Branch is on top of ${GITHUB_BASE_REF}"
+
+        HOOKS_DIR=./git-hooks/hooks GIT_DIR=.git ./git-hooks/hooks/check-commits.sh origin/${GITHUB_BASE_REF}..HEAD^2
```

Comment:
I do not understand the `HEAD^2` .. to me it sounds like it should be just `HEAD^` (as that is the merge commit). Why skip the next commit too?

Reply:
It's a merge commit, so there are more than one parent. `HEAD^` is the first parent, master branch. `HEAD^2` is the second parent, the PR branch.

Reply:
Okay, never knew. Also deserves a comment for sure :D But that means this fails, as the chances are pretty high for a pull-request to not be on `origin/master` (the tip), so this compare will not return any results. I guess that is why the rebase was there :) That somewhat makes sense now ..

You can also use `git merge-base` I guess to find the common point. But for this you need a higher fetch-depth on the `origin/master`, otherwise it won't find them.

All in all, I have to conclude that this really only works for a small subset of our PRs. Most people don't bother updating `master` that often, so pull requests are often behind one or more commits.

I tried to find other ways to do this without fetching the whole repository, but it seems non-trivial. Do you agree with my findings? :)

Reply:
I totally agree :)
I'm also trying to find an elegant way to fetch more only if needed.

Reply:
There are gists that do this; but are they worth it? Complexity vs speed .. :D

I saw some gists where people just did a for-loop increasing `fetch-depth` until `merge-base` returned a result. Silly, but efficient I guess. Just many hits on the server .. not sure GitHub appreciates that :D

Reply:
Yeah, I think I'll start with a reasonable depth (enough for most PRs) and then double the depth in a loop only if it's really needed (for rare huge PRs)